File: src/fzerox_emulator/emulator/race_start.py
# src/fzerox_emulator/emulator/race_start.py
"""Race-start patching methods mixed into the concrete emulator wrapper."""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from fzerox_emulator._native import Emulator as NativeEmulator
    from fzerox_emulator.boundary import RaceStartRequestDict


class RaceStartMixin:
    """Python method surface for native-owned race-start RAM patching."""

    _native: NativeEmulator

    def patch_race_start_setup(
        self,
        *,
        mode: str,
        course_index: int,
        character_index: int,
        engine_setting_raw_value: int,
        total_lap_count: int,
        gp_difficulty_raw_value: int = -1,
    ) -> None:
        """Patch one live race start using native-owned RAM layout rules."""

        self._native.patch_race_start_setup(
            _race_start_request(
                mode=mode,
                course_index=course_index,
                character_index=character_index,
                engine_setting_raw_value=engine_setting_raw_value,
                total_lap_count=total_lap_count,
                gp_difficulty_raw_value=gp_difficulty_raw_value,
            )
        )

    def patch_machine_settings(
        self,
        *,
        mode: str,
        course_index: int,
        character_index: int,
        engine_setting_raw_value: int,
        total_lap_count: int,
        gp_difficulty_raw_value: int = -1,
    ) -> None:
        """Patch menu-level machine settings before race initialization."""

        self._native.patch_machine_settings(
            _race_start_request(
                mode=mode,
                course_index=course_index,
                character_index=character_index,
                engine_setting_raw_value=engine_setting_raw_value,
                total_lap_count=total_lap_count,
                gp_difficulty_raw_value=gp_difficulty_raw_value,
            )
        )

    def patch_engine_settings(
        self,
        *,
        mode: str,
        engine_setting_raw_value: int,
    ) -> None:
        """Patch only engine-related globals for the already selected machine."""

        self._native.patch_engine_settings(mode, engine_setting_raw_value)

    def patch_time_attack_menu_mode(self) -> None:
        """Select the Time Attack branch in the main-menu globals."""

        self._native.patch_time_attack_menu_mode()

    def force_race_reinit(self, *, mode: str) -> None:
        """Force the game to rebuild the current race from menu globals."""

        self._native.force_race_reinit(mode)

    def validate_race_start_setup(
        self,
        *,
        mode: str,
        course_index: int,
        character_index: int,
        engine_setting_raw_value: int,
        total_lap_count: int,
        gp_difficulty_raw_value: int = -1,
    ) -> None:
        """Validate that the native race-start RAM view matches the requested setup."""

        self._native.validate_race_start_setup(
            _race_start_request(
                mode=mode,
                course_index=course_index,
                character_index=character_index,
                engine_setting_raw_value=engine_setting_raw_value,
                total_lap_count=total_lap_count,
                gp_difficulty_raw_value=gp_difficulty_raw_value,
            )
        )

    def vehicle_setup_info(self) -> dict[str, object]:
        """Return native-decoded setup info for HUD/debug checks."""

        return dict(self._native.vehicle_setup_info())


def _race_start_request(
    *,
    mode: str,
    course_index: int,
    character_index: int,
    engine_setting_raw_value: int,
    total_lap_count: int,
    gp_difficulty_raw_value: int,
) -> RaceStartRequestDict:
    return {
        "mode": mode,
        "course_index": course_index,
        "character_index": character_index,
        "machine_skin_index": -1,
        "engine_setting_raw_value": engine_setting_raw_value,
        "total_lap_count": total_lap_count,
        "gp_difficulty_raw_value": gp_difficulty_raw_value,
    }

File: rust/core/host/runtime/race_start.rs
// rust/core/host/runtime/race_start.rs
//! Race-start patching and validation methods on the host runtime.

use crate::core::error::CoreError;
use crate::core::game::race_start::{RaceStartMode, RaceStartSetup, VehicleSetupInfo};

use super::host::Host;

impl Host {
    pub fn patch_time_attack_menu_mode(&mut self) -> Result<(), CoreError> {
        let system_ram = self.system_ram_slice_mut()?;
        crate::core::game::race_start::write_time_attack_menu_mode(system_ram)
    }

    pub fn patch_race_start_setup(
        &mut self,
        mode: RaceStartMode,
        setup: RaceStartSetup,
    ) -> Result<(), CoreError> {
        let system_ram = self.system_ram_slice_mut()?;
        crate::core::game::race_start::write_race_setup(system_ram, mode, setup)
    }

    pub fn patch_machine_settings(
        &mut self,
        mode: RaceStartMode,
        setup: RaceStartSetup,
    ) -> Result<(), CoreError> {
        let system_ram = self.system_ram_slice_mut()?;
        crate::core::game::race_start::write_machine_settings(system_ram, mode, setup)
    }

    pub fn patch_engine_settings(
        &mut self,
        mode: RaceStartMode,
        engine_setting_raw_value: i32,
    ) -> Result<(), CoreError> {
        let system_ram = self.system_ram_slice_mut()?;
        crate::core::game::race_start::write_engine_settings(
            system_ram,
            mode,
            engine_setting_raw_value,
        )
    }

    pub fn force_race_reinit(&mut self, mode: RaceStartMode) -> Result<(), CoreError> {
        let system_ram = self.system_ram_slice_mut()?;
        crate::core::game::race_start::force_race_reinit(system_ram, mode)
    }

    pub fn validate_race_start_setup(
        &mut self,
        mode: RaceStartMode,
        setup: RaceStartSetup,
    ) -> Result<(), CoreError> {
        let system_ram = self.system_ram_slice()?;
        crate::core::game::race_start::validate_race_setup(system_ram, mode, setup)
    }

    pub fn vehicle_setup_info(&mut self) -> Result<VehicleSetupInfo, CoreError> {
        let system_ram = self.system_ram_slice()?;
        crate::core::game::race_start::vehicle_setup_info(system_ram)
    }
}
